common/src/tensorflow/lite/micro/micro_time.cc

```diff
@@ -15,24 +15,36 @@ limitations under the License.
 
 #include "tensorflow/lite/micro/micro_time.h"
 
-static inline unsigned get_mcycle() {
-  unsigned result;
-  asm volatile("csrr %0, mcycle" : "=r"(result));
-  return result;
-}
-
-static inline void set_mcycle(unsigned cyc) {
-  asm volatile("csrw mcycle, %0" ::"r"(cyc));
+// Read the cycle counter.
+//
+// The value of the counter is stored across two 32-bit registers: `mcycle` and
+// `mcycleh`. This function is guaranteed to return a valid 64-bit cycle
+// counter value, even if `mcycle` overflows before reading `mcycleh`.
+//
+// Adapted from: The RISC-V Instruction Set Manual, Volume I: Unprivileged ISA
+// V20191213, pp. 61.
+static inline uint64_t get_mcycle() {
+  uint32_t cycle_low = 0;
+  uint32_t cycle_high = 0;
+  uint32_t cycle_high_2 = 0;
+  asm volatile(
+      "read%=:"
+      "  csrr %0, mcycleh;"     // Read `mcycleh`.
+      "  csrr %1, mcycle;"      // Read `mcycle`.
+      "  csrr %2, mcycleh;"     // Read `mcycleh` again.
+      "  bne  %0, %2, read%=;"  // Try again if `mcycle` overflowed before
+                                // reading `mcycleh`.
+      : "+r"(cycle_high), "=r"(cycle_low), "+r"(cycle_high_2)
+      :);
+  return (uint64_t) cycle_high << 32 | cycle_low;
 }
 
 namespace tflite {
 
 // Assumes 100 MHz
-int32_t ticks_per_second() { return 100000000; }
+int32_t ticks_per_second() { return 100000000 / 1024; }
```

Comment:
Passing comment, not necessary to make a change in this PR: We ought to be able to use CONFIG_CLOCK_FREQUENCY from generated/soc.h here, right?

Reply:
You're right, this is Arty-specific and will be wrong for many other boards.